libs/vkd3d/memory.c

```diff
@@ -411,18 +616,128 @@ void vkd3d_memory_allocator_cleanup(struct vkd3d_memory_allocator *allocator, st
     pthread_mutex_destroy(&allocator->mutex);
 }
 
+static HRESULT vkd3d_memory_allocator_add_chunk(struct vkd3d_memory_allocator *allocator, struct d3d12_device *device,
+        const D3D12_HEAP_PROPERTIES *heap_properties, D3D12_HEAP_FLAGS heap_flags, uint32_t type_mask, struct vkd3d_memory_chunk **chunk)
+{
+    struct vkd3d_allocate_memory_info alloc_info;
+    struct vkd3d_memory_chunk *object;
+    HRESULT hr;
+
+    memset(&alloc_info, 0, sizeof(alloc_info));
+    alloc_info.memory_requirements.size = VKD3D_MEMORY_CHUNK_SIZE;
+    alloc_info.memory_requirements.alignment = 0;
+    alloc_info.memory_requirements.memoryTypeBits = type_mask;
+    alloc_info.heap_properties = *heap_properties;
+    alloc_info.heap_flags = heap_flags;
+
+    if (!(heap_flags & D3D12_HEAP_FLAG_DENY_BUFFERS))
+        alloc_info.flags |= VKD3D_ALLOCATION_FLAG_GLOBAL_BUFFER;
+
+    if (!vkd3d_array_reserve((void**)&allocator->chunks, &allocator->chunks_size,
+            allocator->chunks_count + 1, sizeof(*allocator->chunks)))
+    {
+        ERR("Failed to allocate space for new chunk.\n");
+        return E_OUTOFMEMORY;
+    }
+
+    if (FAILED(hr = vkd3d_memory_chunk_create(device, allocator, &alloc_info, &object)))
+        return hr;
+
+    allocator->chunks[allocator->chunks_count++] = *chunk = object;
+    return S_OK;
+}
+
+static HRESULT vkd3d_memory_allocator_try_suballocate_memory(struct vkd3d_memory_allocator *allocator,
+        struct d3d12_device *device, const VkMemoryRequirements *memory_requirements, uint32_t type_mask,
+        const D3D12_HEAP_PROPERTIES *heap_properties, D3D12_HEAP_FLAGS heap_flags,
+        struct vkd3d_memory_allocation *allocation)
+{
+    struct vkd3d_memory_chunk *chunk;
+    HRESULT hr;
+    size_t i;
+
+    type_mask &= device->memory_info.global_mask;
+    type_mask &= memory_requirements->memoryTypeBits;
+
+    for (i = 0; i < allocator->chunks_count; i++)
+    {
+        chunk = allocator->chunks[i];
+
+        /* Match flags since otherwise the backing buffer
+         * may not support our required usage flags */
+        if (chunk->allocation.d3d12_flags != heap_flags)
```

Comment:
Will this match CUSTOM properties as well? I.e. CACHED vs UNCACHED CUSTOM heap type.

Reply:
Custom properties shouldn't matter at all since they only affect the memory type, not the usage flags.

I suppose we have to match the heap type though, since allocations on different heap types can target the same memory type and that does actually affect usage flags.